lib: mempool: Fix issues in mempool allocator

pdunaj opened the pull request:
This commits fixes three issues:
Fixes: #14504
1. Allocator returning success even if allocation failed

Fixes: #14508
2. Allocator fails to allocate while memory is available
but was hidden by other context calling free.
3. Allocator fails to allocate while memory is available
but was moved to another level.

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

andyross commented on lib/os/mempool.c:
```diff
@@ -236,67 +231,72 @@ static void *block_break(struct sys_mem_pool_base *p, void *block, int l,
+	unsigned int key = pool_irq_lock(p);
+
+	/* Try allocating the smallest block possible.
+	 * Search must be performed atomically to make sure block rearranging
```
The existing design was intended to prevent exactly this circumstance, though.  Making the allocation fully atomic is a latency bug.

pdunaj replied:
The only thing added is checking if level is empty, these are just few operations. Note that `block_alloc` will return immediately if level is empty. Only when there is block available on a level checked `block_alloc` will do something expensive.
So you will end up with once expensive `block_alloc` and few `block_break` (if needed) - both already under a lock.
I personally prefer that to unbound operation with repetitive retries.

andyross reviewed (changes requested):
I'm staring at this, and am not entirely convinced this is correct.  Can you share the exerciser code that triggers the issue in the linked bugs?  I think there may be something else going on.

The issue is for sure subtle, but we've been through here several times now and I'm almost certain the current design is correct.

andyross commented:
(Sorry, put this comment in the wrong spot earlier):

Yeah, we can't do it this way. The fix effectively wraps the whole allocation in a lock, as I understand it. That's counter to the original design intent. Maybe try flagging the existence of a race somehow with a "last thread to enter allocator" state variable or something as a cue to retry?

pdunaj commented:
Hi @andyross , if you check again you will notice that the amount of the code under the lock is not really increased. What was added is (essentially) checking if free list on n level is empty. And you need to do it under a lock anyway as otherwise content of the list may be broken. Costly operations of alloc and break are already under a lock.

Yes, there is an alternative. Each time allocator state is updated a token could be changed. Block allocator could take a token, do barrier, try allocating and in case there is nomem, grab the current token and if it changed do a retry. I was thinking about this solution but the main problem is that it is unbound. Second thing is that, as I written above, the change is bringing under a lock only checks which levels are empty (which is really cheap operation). The third thing is that such solutions are prone to bugs.

The current state is away from the original design already as locks were introduced to handle various races. Please review the code again - the number of operations added under the lock is really small.

andyross commented on lib/os/mempool.c:
```diff
@@ -236,67 +231,72 @@ static void *block_break(struct sys_mem_pool_base *p, void *block, int l,
+	 */
+	int from_l;
+	for (from_l = alloc_l; from_l >= 0; from_l--) {
+		data = block_alloc(p, from_l, lsizes[from_l]);
```
This is the latency bug: the old code was lockless here, but the new code can potentially call block_alloc() once for every level in the allocator before releasing the lock (for reference, a 32k heap of 16 byte blocks has 8 levels).  That's a lot of cycles added to worst case latency!

pdunaj replied:
I have made some measurements today to check the impact.
I write to GPIO just after lock is taken and just before lock released (not including the internal release/lock done on break). This should measure worst case scenario under the lock and show the additional time spent.

I noticed that previous organization of code gave slightly better timing so I stick with it (comments are not updated yet as there is no point if change is not approved).

I used logic analyzer with 25 MS/s so resolution should be ok.
Testing code does `k_malloc`/`k_free` of given size. As nothing is allocated from the heap before the test this is the worst case scenario where blocks will have to be broken down on higher levels.
Heap was configured as follows: 64B - 1024B (default config in our project). I can reconfigure it to have more levels but the pattern is visible so there may be no point.

This was performed on board with nrf52840.

```
size (B) |  base (us)  | change (us)
-------- | ----------- | -----------
      16 | 8.12 - 8.64 | 8.36 - 8.92
      64 | 5.00 - 5.52 | 5.12 - 5.72
     512 | 1.72 - 2.00 us | 1.72 - 2.08
```

The impact seems to be:
- none if block is available on level
- around 0.12 us per level (block alloc call finds level empty)

Cost of 2x gpio write should be around 0.8 us or so.

Cost of block allocation: around 1 us
Cost of single block break: around 3.2 us

I noticed that current code relax the lock after `block_break` which makes latency sum of both times. I moved the relax before break so now the maximum latency is max of block alloc and break times.
In proposed change we have N checks on level being empty added to normal `block_alloc` time. To have max latency on the same level as `block_break` number of levels would have to be around `(3.2 - 1)/0.12` that is around 18.

pdunaj replied:
I have increased heap and checked if timings are as expected on higher levels.
```
   level |   base (us)  |  change (us)
-------- | ------------ | ------------
       4 | 14.32 - 14.88 | 14.8 - 15.24
       3 | 11.2 - 11.76 | 11.6 -  12.04
       2 |  8.12 -  8.64 |  8.36 -  8.92
       1 |  5.00 -  5.52 |  5.12 -  5.72
       0 |  1.72 -  2.00 |  1.72 -  2.08
```
It seems to be scaling as expected.

pdunaj replied:
Hi @andyross , have you had a chance to look at the values? The worst case scenario is not increased and I would prefer to leave implementation as is to avoid further problems in this area.
If you are fine I will update the comments.

pdunaj committed:
lib: mempool: Synchronize level checks

Do not perform early level usage check. This can lead to situation
where block is seen as available on level when it was taken from
the other context.

Fixes: #14504

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

pdunaj committed:
lib: mempool: Return error if no block found

Return -ENOMEM if no block is available on any level.

Fixes: #14504

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

pdunaj commented:
FYI @pizi-nordic

andyross reviewed (approved):
OK, the holding-the-lock-around-all-the-levels behavior that I thought was there isn't (or maybe you fixed it?).  This design still relaxes the lock at each level, so it seems clean to me.  As expressed here it looks ismorphic to the previous behavior, but with a much smaller window with which to race against other allocators (as I see it it's still possible for a racing free to land "below" the one here and technically add a block that could be returned, but as mentioned I'm not personally convinced that racy "busy memory" is a bug per se).

I'm cool with this.  Thanks.

pdunaj commented:
thanks
